operators/src/processing/new_expression/mod.rs

```diff
@@ -0,0 +1,749 @@
+/// * `output_no_data_value` is the no data value of the output raster
+/// * `output_measurement` is the measurement description of the output
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct ExpressionParams {
+    pub expression: String,
+    pub output_type: RasterDataType,
+    #[serde(with = "float_with_nan")]
+    pub output_no_data_value: f64, // TODO: check value is valid for given output type during deserialization
+    pub output_measurement: Option<Measurement>,
+    pub map_no_data: bool,
+}
+
+// TODO: rename to `Expression`
+/// The `Expression` operator calculates an expression for all pixels of the input rasters and
+/// produces raster tiles of a given output type
+pub type NewExpression = Operator<ExpressionParams, ExpressionSources>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ExpressionSources {
+    a: Box<dyn RasterOperator>,
+    b: Option<Box<dyn RasterOperator>>,
+    c: Option<Box<dyn RasterOperator>>,
+}
+
+impl OperatorDatasets for ExpressionSources {
+    fn datasets_collect(&self, datasets: &mut Vec<DatasetId>) {
+        self.a.datasets_collect(datasets);
+
+        if let Some(ref b) = self.b {
+            b.datasets_collect(datasets);
+        }
+
+        if let Some(ref c) = self.c {
+            c.datasets_collect(datasets);
+        }
+    }
+}
+
+impl ExpressionSources {
+    pub fn new_a(a: Box<dyn RasterOperator>) -> Self {
+        Self {
+            a,
+            b: None,
+            c: None,
+        }
+    }
+
+    pub fn new_a_b(a: Box<dyn RasterOperator>, b: Box<dyn RasterOperator>) -> Self {
+        Self {
+            a,
+            b: Some(b),
+            c: None,
+        }
+    }
+
+    pub fn new_a_b_c(
+        a: Box<dyn RasterOperator>,
+        b: Box<dyn RasterOperator>,
+        c: Box<dyn RasterOperator>,
+    ) -> Self {
+        Self {
+            a,
+            b: Some(b),
+            c: Some(c),
+        }
+    }
+
+    fn number_of_sources(&self) -> usize {
+        let a: usize = 1;
+        let b: usize = self.b.is_some().into();
+        let c: usize = self.c.is_some().into();
+
+        a + b + c
+    }
+
+    async fn initialize(
+        self,
+        context: &dyn ExecutionContext,
+    ) -> Result<ExpressionInitializedSources> {
+        let b = if let Some(b) = self.b {
+            Some(b.initialize(context).await)
+        } else {
+            None
+        };
+
+        let c = if let Some(c) = self.c {
+            Some(c.initialize(context).await)
+        } else {
+            None
+        };
+
+        Ok(ExpressionInitializedSources {
+            a: self.a.initialize(context).await?,
+            b: b.transpose()?,
+            c: c.transpose()?,
+        })
+    }
+}
+
+/// Create a parameter name from an index.
+/// Starts with `A`.
+fn index_to_parameter(index: usize) -> String {
+    // TODO: handle indices > 26
```

Comment:
why? the source only supports three inputs

Reply:
084ff66

Reply:
I don't see this addressed in the commit

Reply:
a74a07c